### PR TITLE
[MU3] Fix #127341: Implements portamento for fluid

### PR DESCRIPTION
**Pull request by Jojo-Schmitz**
to replace/rebase @MichaelFroelich's #4824

Resolves https://musescore.org/en/node/127341

### CODE REVIEW

File: fluid/fluid.cpp [hunk 1]
```diff
@@ -38,6 +38,11 @@ namespace FluidS {
 
 bool Fluid::initialized = false;
 
+/* better than a macro to determine inappropriate values for notes*/
+bool validNote(const int input) {
+      return (input < 255 && input > -1);
```
Comment (Marr11317) [inline fluid/fluid.cpp]: shouldn't it be `input < 256` or `input <= 255`?

Comment (Jojo-Schmitz) [reply fluid/fluid.cpp]: Good question. At another place in this file this is claimed to "usually be the TPC" (Tonal Pitch Class), but that  in turn has a range of -2 (`TPC_INVALID`) to 33 (`TPC_B_SS`), so that doesn't seem to be the case.

Comment (Jojo-Schmitz) [reply fluid/fluid.cpp]: It it were pitches though, the limit would rather be 0-127

Comment (Jojo-Schmitz) [reply fluid/fluid.cpp]: However, as far as I can tell this `validNote()` is call only from `Fluid::setFromKeyPortamento()` and that in turn is never called!?!

Comment (MichaelFroelich) [reply fluid/fluid.cpp]: This code was copied from fluid synth 2, which isn't quite possible to completely incorporate into MuseScore. Jojo could actually be right, I just wasn't entirely too sure what's necessary for portamento and what's not, or what might be necessary in the future, and I wanted to add as much of fluid synth 2's extra features as possible. It's specifically to do with line 219 of this file:

https://github.com/FluidSynth/fluidsynth/blob/master/src/synth/fluid_chan.h#L219

If the file has changed, look for a macro like `#define fluid_channel_is_valid_note(n)    (n != INVALID_NOTE)` I'm not a fan of macros because I think they can occasionally be unpredictable and otherwise some compilers may or may not optimise them well or add unnecessary symbols.

Long story short, 255 is indeed an invalid value as is any value that's negative.

File: libmscore/rendermidi.cpp [hunk 2]
```diff
@@ -238,6 +248,45 @@ static void playNote(EventMap* events, const Note* note, int channel, int pitch,
...
+                        portamentoControl.setOriginatingStaff(staffIdx);
+                        NPlayEvent portamentoTimeMSB(ME_CONTROLLER, channel, CTRL_PORTAMENTO_TIME_MSB, msb);
+                        portamentoTimeMSB.setOriginatingStaff(staffIdx);
+                        NPlayEvent portamentoTimeLSB(ME_CONTROLLER, channel, CTRL_PORTAMENTO_TIME_LSB, lsb);
+                        portamentoTimeLSB.setOriginatingStaff(staffIdx);
+                        
+                        ev.setPortamento(true);
+                        if (onTime == 0)
+                              onTime++;
+                        events->insert(std::pair<int, NPlayEvent>(onTime-1, portamentoOn));
```
Comment (MichaelFroelich) [inline libmscore/rendermidi.cpp]: Would these events also be present if I exported a midi? That's kind of where I got stuck.

File: fluid/fluid.cpp (hunk 1 above)
Comment (Jojo-Schmitz) [reply fluid/fluid.cpp]: * So `const int INVALID_NOTE = 255;` rather than a magic number?
Instead with have this though:

      enum { INVALID_NOTE = -1};

Confusing? Yes...

* That comment about TPC seems missleading maybe it means PTC, PortamenTo Controler? Or is it really TPC?

File: libmscore/rendermidi.cpp (hunk 2 above)
Comment (Jojo-Schmitz) [reply libmscore/rendermidi.cpp]: No idea. How to check?

Anyway, are you going to continue on your PR then? The rebase itself was rather trivial.

Comment (Jojo-Schmitz) [reply libmscore/rendermidi.cpp]: Actually I see you did rebase your PR, so I'm going to close mine

**Review by igorkorsukov — APPROVED** [review]
